File: MobMuPlatEditor-OSXLionAndUp/MobMuPlatEditor/Controls/MMPControl.m
//
//  MMPControl.m
//  MobMuPlatEd1
//
//  Created by Daniel Iglesia on 12/26/12.
//  Copyright (c) 2012 Daniel Iglesia. All rights reserved.
//
//  Abstract superclass of all gui control widget classes.


#import "MMPControl.h"
#define HANDLE_SIZE 20 //size of the EditHandle in the lower right corner
#define PASTE_OFFSET 20 //when a MMPControl gets copeid and pasted, offset it it this much (to the right and down)
#define BORDER_WIDTH 5 //editing border
#import "CanvasView.h"

@implementation MMPControl
@synthesize isSelected, editingDelegate;

- (id)initWithFrame:(NSRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        [self setWantsLayer:YES];
        //default colors!
        [self setColor:[NSColor blueColor]];
        [self setHighlightColor:[NSColor redColor]];
    }
    return self;
}

//coordinate system is flipped upside down, so that it matches the coordinate system of iOS
-(BOOL) isFlipped{
  return YES;
}

//put the EditHandle on the lower right when editing
-(void)addHandles{
  _handle = [[EditHandle alloc]initWithFrame:CGRectMake(self.frame.size.width-HANDLE_SIZE, self.frame.size.height-HANDLE_SIZE, HANDLE_SIZE, HANDLE_SIZE) ];

  _handle.hidden=YES;
  [self addSubview:_handle];
}

-(void)hackRefresh{
 if(isSelected){
   [self.layer setBorderWidth:BORDER_WIDTH];
 }
 [self setColor:[self color]];
 [self bringUpHandle];
 }

-(void)bringUpHandle{//bringSubviewToFront doesn't seem to work?
  [_handle removeFromSuperview];
  [self addSubview:_handle];
  _handle.layer.borderWidth=BORDER_WIDTH;
}

-(void)mouseDown:(NSEvent *)event{

  if([editingDelegate isEditing]){

    //SELECTION
    wasSelectedThisCycle=NO;
    BOOL wasAlreadySelected = isSelected;
    if(!isSelected){
      [self setIsSelected:YES];//toggle
      wasSelectedThisCycle=YES;
    }

    [editingDelegate controlEditClicked:self withShift:([event modifierFlags] & NSShiftKeyMask)!=0 wasAlreadySelected:wasAlreadySelected ];//poll objects for selection state

    //MOVEMENT
    lastDragLocation=[[self superview] convertPoint:[event locationInWindow]fromView:nil];
    clickOffsetInMe=[self convertPoint:[event locationInWindow] fromView:nil];

    [self.editingDelegate updateGuide:self]; //TODO move to delegate controlEditClicked
  }
}

-(void)setFrameObjectUndoable:(NSValue*)frameObject{
  [[self undoManager] registerUndoWithTarget:self selector:@selector(setFrameObjectUndoable:) object:[NSValue valueWithRect:self.frame]];
  [self setFrame:[frameObject rectValue]];
}

-(void)setFrame:(NSRect)frameRect{
  [super setFrame:frameRect];
  if(_handle){
    [_handle setFrameOrigin:CGPointMake(frameRect.size.width-HANDLE_SIZE,self.frame.size.height-HANDLE_SIZE)];
  }
}

-(void)setFrameOriginObjectUndoable:(NSValue*)pointObject{
  [[self undoManager] registerUndoWithTarget:self selector:@selector(setFrameOriginObjectUndoable:) object:[NSValue valueWithPoint:self.frame.origin]];
  [self setFrameOrigin:[pointObject pointValue]];
}

-(void)mouseDragged:(NSEvent *)event{

  if([self.editingDelegate isEditing]){
    if(dragging==NO){//first drag
      [[self undoManager] beginUndoGrouping];
    }
    dragging=YES;

    [[NSCursor closedHandCursor] push];
    NSPoint newDragLocation=[[self superview] convertPoint:[event locationInWindow] fromView:nil];
    //CGPoint newOrigin = CGPointMake(newDragLocation.x-clickOffsetInMe.x, newDragLocation.y-clickOffsetInMe.y);

    //[self setFrameOriginObjectUndoable:[NSValue valueWithPoint:newOrigin]]; //move movement to delegate
    [self autoscroll:event];
    [self.editingDelegate controlEditMoved:self deltaPoint:CGPointMake(newDragLocation.x-lastDragLocation.x, newDragLocation.y-lastDragLocation.y)];
    //[self.editingDelegate updateGuide:self];
    lastDragLocation=newDragLocation;
  }
}

-(void)mouseUp:(NSEvent *)theEvent  {

  if([self.editingDelegate isEditing]){
    //if end of dragging
    if(dragging){
      [[self  undoManager] endUndoGrouping];
      [[NSCursor closedHandCursor] pop];
      dragging=NO;
      [self.editingDelegate controlEditReleased:self withShift:([theEvent modifierFlags] & NSShiftKeyMask)!=0 hadDrag:YES];//don't think we need this for anything...
    }
    //wasn't a drag, just a click+release, and if has shift
    else{
      if(([theEvent modifierFlags] & NSShiftKeyMask)!=0 && isSelected && !wasSelectedThisCycle) [self setIsSelected:NO];//allow shift toggle off
      [self.editingDelegate controlEditReleased:self withShift:([theEvent modifierFlags] & NSShiftKeyMask)!=0 hadDrag:NO];
    }
    [self.editingDelegate updateGuide:nil];
  }
}



-(void)setIsSelected:(BOOL)newSelected{

  isSelected=newSelected;
  if(isSelected){
    [self.layer setBorderWidth:BORDER_WIDTH];
    [_handle setHidden:NO];
  }
  else{
    [self.layer setBorderWidth:0];
    [_handle setHidden:YES];
  }
}

/*+(CGColorRef) CGColorFromNSColor:(NSColor*)inColor{

 CGFloat components[4];
 [inColor getComponents:components];
 return CGColorCreateGenericRGB(components[0], components[1], components[2], components[3]);
 }*/

//====copy/paste
- (void)encodeWithCoder:(NSCoder *)coder {
  [super encodeWithCoder:coder];
  [coder encodeObject:NSStringFromRect(CGRectMake(self.frame.origin.x+PASTE_OFFSET, self.frame.origin.y+PASTE_OFFSET, self.frame.size.width, self.frame.size.height)) forKey:@"frame"];//string->rect
  [coder encodeObject:self.address forKey:@"address"];//string
  [coder encodeObject:self.color forKey:@"color"];//color
  [coder encodeObject:self.highlightColor forKey:@"highlightColor"];//color
  //[coder encodeObject:self.handle forKey:@"handle"];

}

- (id)initWithCoder:(NSCoder *)coder {
  CGRect frame = NSRectFromString([coder decodeObjectForKey:@"frame"]);
  self = [[[self class] alloc] initWithFrame:frame];
  if(self){//=[super initWithCoder:coder]){
    //[self setWantsLayer:YES];
    //[self setFrame:NSRectFromString([coder decodeObjectForKey:@"frame"])];
    self.address = [coder decodeObjectForKey:@"address"];
    [self setColor:[coder decodeObjectForKey:@"color"]];
    [self setHighlightColor:[coder decodeObjectForKey:@"highlightColor"]];
  }
  return self;
}

- (NSArray *)writableTypesForPasteboard:(NSPasteboard *)pasteboard {
  return [NSArray arrayWithObject:@"com.iglesiaintermedia.mobmuplat.mmpcontrol"];
}

+ (NSArray *)readableTypesForPasteboard:(NSPasteboard *)pasteboard {
  return [NSArray arrayWithObject:@"com.iglesiaintermedia.mobmuplat.mmpcontrol"];
}

+ (NSPasteboardReadingOptions)readingOptionsForType:(NSString *)type pasteboard:(NSPasteboard *)pboard {
  return NSPasteboardReadingAsKeyedArchive;
}

- (id)pasteboardPropertyListForType:(NSString *)type {
  return [NSKeyedArchiver archivedDataWithRootObject:self];
}


-(BOOL)canBecomeFirstResponder{
  return YES;
}

-(void)keyDown:(NSEvent*)event{
  [self interpretKeyEvents:[NSArray arrayWithObject:event]];
}

//responder chain for keyboard "delete" hit
-(IBAction)deleteBackward:(id)sender{
  //printf("\ndel!");
  [self.editingDelegate controlEditDelete];
}

//more undoables

-(void)setAddressUndoable:(NSString *)address{
  [[self undoManager] registerUndoWithTarget:self selector:@selector(setAddressUndoable:)  object:self.address];
  [self setAddress:address];
}

-(void)setColorUndoable:(NSColor *)color{
  [[self undoManager] registerUndoWithTarget:self selector:@selector(setColorUndoable:) object:self.color];
  [self setColor:color];
}

-(void)setHighlightColorUndoable:(NSColor *)color{
  [[self undoManager] registerUndoWithTarget:self selector:@selector(setHighlightColorUndoable:) object:self.highlightColor];
  [self setHighlightColor:color];
}

//all subclasses override this when receiveing a message
-(void)receiveList:(NSArray *)inArray{
  if ([inArray count] >= 2 &&
      [inArray[0] isKindOfClass:[NSString class]] &&
      [inArray[0] isEqualToString:@"enable"] &&
      [inArray[1] isKindOfClass:[NSNumber class]]) {
    self.enabled = ([inArray[1] floatValue] > 0);
  }
}

- (void)setEnabled:(BOOL)enabled {
  [super setEnabled:enabled];
  self.alphaValue = enabled ? 1 : .2;
}

@end

File: MobMuPlatEditor-Java/src/com/iglesiaintermedia/MobMuPlatEditor/controls/MMPControl.java
package com.iglesiaintermedia.MobMuPlatEditor.controls;

import java.awt.Color;
import java.awt.Rectangle;
import java.awt.Point;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.net.InetAddress;
import java.net.SocketException;
import java.net.UnknownHostException;
import java.util.ArrayList;

import com.iglesiaintermedia.MobMuPlatEditor.*;

import javax.swing.*;
import javax.swing.SwingUtilities.*;
import javax.swing.border.Border;

import com.illposed.osc.*;

public class MMPControl extends JPanel implements MouseListener, MouseMotionListener{
	final static int HANDLE_SIZE = 20; //size of the EditHandle in the lower right corner
	
	String address;
	private Color _color, _disabledColor;
	private Color _highlightColor, _disabledHighlightColor;
	public boolean isSelected;
	public MMPController editingDelegate;
	
	private  boolean dragging;
	private Point clickOffsetInMe;
	private Point lastDragLocation;
	private boolean wasSelectedThisCycle;
	EditHandle handle;
	
	protected Border border = BorderFactory.createLineBorder(Color.black, 5);
	JPanel topPanel;
	
	
		
	public MMPControl(){
		super();
		setLayout(null);
		setOpaque(false);
		_color = Color.BLUE;//new Color(1f,.1f,.1f,.1f);
		_highlightColor = Color.RED;//new Color(0f,.1f,.1f,.1f);
		
		topPanel = new JPanel();
		topPanel.setOpaque(false);
		add(topPanel);
		
		addHandles();//do this first in java, instead of last in osx
		
	}
	
	public void setAddress(String newAddress){
		address = newAddress;
	}
	public String getAddress() {
		return address;
	}
	
	public void setIsSelected(boolean inIsSelected){
		isSelected=inIsSelected;
		if(isSelected){
			topPanel.setBorder(border);// [self.layer setBorderWidth:5];
        	handle.setVisible(true);
    	}
    	else{
    		topPanel.setBorder(null);//orsetBorderPainted(false)//[self.layer setBorderWidth:0];
        	handle.setVisible(false);
    	}
	
	}
	
	public void addHandles(){
		  handle = new EditHandle();
	      handle.setBounds(getBounds().width-HANDLE_SIZE,getBounds().height-HANDLE_SIZE, HANDLE_SIZE, HANDLE_SIZE);
	    handle.setVisible(false);
	   add(handle);
	} 

	public void setBounds(Rectangle bounds){
		super.setBounds(bounds);
		topPanel.setBounds(0,0,bounds.width, bounds.height);
		if(handle!=null)handle.setBounds(getBounds().width-HANDLE_SIZE,getBounds().height-HANDLE_SIZE, HANDLE_SIZE, HANDLE_SIZE);
	}
	
	public void setColor(Color newColor){
		_color = newColor;
		_disabledColor = new Color(newColor.getRed(), newColor.getGreen(), newColor.getBlue(), (int)(newColor.getAlpha() * .2));
	}
	
	public Color getColor() {
		return  _color;
	}
	
	public void setHighlightColor(Color newColor){
		_highlightColor = newColor;
		_disabledHighlightColor = new Color(newColor.getRed(), newColor.getGreen(), newColor.getBlue(), (int)(newColor.getAlpha() * .2));
	}
	
	public Color getHighlightColor() {
		return _highlightColor;
	}
	
	public Color getDisabledColor() {
		return _disabledColor;
	}
	
	public Color getDisabledHighlightColor() {
		return _disabledHighlightColor;
	}
	
	
	public void receiveList(ArrayList<Object> messageArray){
		if (messageArray.size()>=2 && 
				(messageArray.get(0) instanceof String) && 
				messageArray.get(0).equals("enable") && 
				(messageArray.get(1) instanceof Float)) {
			boolean enabled = ((Float)(messageArray.get(1))).floatValue() > 0;
			this.setEnabled(enabled);
			//this.setEnabled(((Float)(messageArray.get(1))).floatValue() > 0);
			//this.setAlpha(this.isEnabled() ? 1.0f : .2f );
			//this.setBackground(this.isEnabled() ? new Color(128, 128 , 128, 128) : new Color(128, 128, 128, 0));
			//this.setColor(new Color(255, 0 , 0, 128));
			//this.setBackground(new Color(128, 0 , 255, 128));
			
		}
	}

	@Override
	public void mouseClicked(MouseEvent e) {//just selection, no motion or interaction
		//calls pressed and released
	}

	@Override
	public void mouseEntered(MouseEvent arg0) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void mouseExited(MouseEvent arg0) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void mousePressed(MouseEvent e) {
		//System.out.print("\nmp "+this.address);
		if(editingDelegate.isEditing()){
			wasSelectedThisCycle=false;
	        boolean wasAlreadySelected = isSelected;
	        if(!isSelected){
	        	this.setIsSelected(true);//toggle
	            wasSelectedThisCycle=true;
	        }
	        
	        //System.out.print("\ncontrolclick isshift "+e.isShiftDown()+" was "+wasAlreadySelected);
			editingDelegate.controlEditClicked(this, e.isShiftDown(), wasAlreadySelected); 
	        //MOVEMENT
	        lastDragLocation=SwingUtilities.convertPoint(this, e.getPoint(), this.getParent());//e.getPoint();//[[self superview] convertPoint:[event locationInWindow]fromView:nil];
	        clickOffsetInMe=e.getPoint();//[self convertPoint:[event locationInWindow] fromView:nil];
	        //System.out.println("lastdrag in canvas "+lastDragLocation.getX()+" "+lastDragLocation.getY()+" offsetinwidget "+clickOffsetInMe.getX()+" "+clickOffsetInMe.getY());
	        editingDelegate.updateGuide(this);
		}
	}

	@Override
	public void mouseReleased(MouseEvent e) {
		//System.out.print("mr");
		if(editingDelegate.isEditing()){
			if(dragging){
				dragging = false;
				editingDelegate.controlEditReleased(this, e.isShiftDown(), true);
			}
			//wasn't a drag, just a click+release, and if has shift
			else{
				if( e.isShiftDown() && isSelected && !wasSelectedThisCycle)setIsSelected(false);//allow shift toggle off
				editingDelegate.controlEditReleased( this, e.isShiftDown(), false);//:self withShift:([theEvent modifierFlags] & NSShiftKeyMask)!=0 hadDrag:NO];
				
			}
			editingDelegate.updateGuide(null);
		}
	}

	@Override
	public void mouseDragged(MouseEvent e) {
		if(editingDelegate.isEditing()){
	        if(dragging==false){//first drag
	           // [[self undoManager] beginUndoGrouping];
	        }
	        dragging=true;
	        
	       // [[NSCursor closedHandCursor] push];
	        Point newDragLocation=SwingUtilities.convertPoint(this, e.getPoint(), this.getParent());//[[self superview] convertPoint:[event locationInWindow] fromView:nil];
	  //      Point newOrigin = new Point((int)(newDragLocation.getX()-clickOffsetInMe.getX()), (int)(newDragLocation.getY()-clickOffsetInMe.getY()));//CGPointMake(newDragLocation.x-clickOffsetInMe.x, newDragLocation.y-clickOffsetInMe.y);
	        
	    //    this.setLocation(newOrigin);//[self setFrameOriginObjectUndoable:[NSValue valueWithPoint:newOrigin]];
	        
	        editingDelegate.controlEditMoved(this, new Point(newDragLocation.x-lastDragLocation.x, newDragLocation.y-lastDragLocation.y));
	        editingDelegate.updateGuide(this);
	        lastDragLocation=newDragLocation;
	    }	
	}

	@Override
	public void mouseMoved(MouseEvent arg0) {
		// TODO Auto-generated method stub
		
	}
	
	// Ignore and pass through mouse events if not editing and widget is disabled.
	public boolean contains(int x, int y) {
		if (!this.isEnabled() && !editingDelegate.isEditing()) {
			return false;
		}
		return super.contains(x, y);
	}
}
